### ISSUE TITLE
emrun always warns the browser is detached and not able to kill browser process at exit

### ISSUE THREAD

**Issue by weitjong** [kripken/emscripten#3234]
I am using CTest (from CMake) to invoke emrun to serve my test html which has an internal logic to call exit() when the time is up. My test setup almost work. The only problem is, emrun fails to kill the browser process at exit before it is about to end. The CTest somehow detects that one of the child process (browser process) spawned by the emrun is still running although emrun itself is already ended; and CTest keeps waiting for it instead of moving on to other test cases. Manually closing the browser windows does the trick. However, this is hardly a solution as I have quite a number of test cases to run non-interactively.

Why is the browser process is always detached? I have tested using FF on Linux and also FF on Mac. Both gets this warning.

**Comment by juj** [comment]
I suppose you are using the `--kill_exit` command line option? How does your full emrun cmdline look like?

**Comment by weitjong** [comment]
Thanks for your reply. Yes, I am using that option. The command is generated by our CMake macro. It looks something like this: 
+ add_test (NAME ${ARG_NAME} COMMAND ${EMRUN} --kill_exit ${TARGET_NAME}.html ${ARG_OPTIONS} WORKING_DIRECTORY ${CMAKE_RUNTIME_OUTPUT_DIRECTORY})

You can also see this line of code in the commit referencing this issue in Github.

**Comment by juj** [comment]
If you run the code in a standalone manner outside the CMake test harness, does it clean up properly? Also, if you run with `--verbose`, do you get any more debug info?

There is one fix to emrun I haven't had time to put in yet, see here: https://github.com/juj/emscripten/commit/fb3b4b1d986442354b5e8c0bbc88ed2a2f3d169e . That fixes an issue that emrun would not be able to kill a browser if the page execution times out. If you have a second, can you try that patch out?

**Comment by weitjong** [comment]
> If you run the code in a standalone manner outside the CMake test harness, does it clean up properly? Also, if you run with --verbose, do you get any more debug info?

When the html is built without ```--emrun``` flag then emrun does not return the control back to CLI regardless (even after manually closing the browser window). The only way to gain back the control is via Ctrl+C. I guess this is somewhat expected.

When the html is built with ```--emrun``` flag then emrun is able to gain back the control after the compiled code calling the ```emscripten_force_exit(EXIT_SUCCESS)```. However, the browser window (i.e. the browser process) remains even when the emrun is invoked with ```--kill_exit```. And with ```--verbose``` I got this:
```
Web server root directory: /home/weitjong/SDKs/urho3d/emscripten-Build/bin
Starting web server in port 6931
Executing /usr/bin/xdg-open http://localhost:6931/13_Ragdolls.html
Warning: emrun got detached from the target browser process. Cannot detect when user closes the browser. Behaving as if --serve_after_close was passed in.
pre-main prep time: 58 ms
[Thu Mar 12 21:13:27 2015] WARNING: Could not get application preferences directory
Calling stub instead of sigaction()
Calling stub instead of sigaction()
[Thu Mar 12 21:13:27 2015] INFO: Opened log file Ragdolls.log
[Thu Mar 12 21:13:27 2015] INFO: Added resource package Data.pak
[Thu Mar 12 21:13:27 2015] INFO: Added resource package CoreData.pak
[Thu Mar 12 21:13:27 2015] INFO: Set screen mode 1024x768 windowed
[Thu Mar 12 21:13:27 2015] INFO: Initialized input
[Thu Mar 12 21:13:27 2015] INFO: Initialized user interface
[Thu Mar 12 21:13:27 2015] INFO: Initialized renderer
[Thu Mar 12 21:13:27 2015] INFO: Set audio mode 44100 Hz stereo interpolated
[Thu Mar 12 21:13:27 2015] INFO: Initialized engine
Web page has quit with a call to exit() with return code 0. Shutting down web server. Pass --serve_after_exit to keep serving even after the page terminates with exit().
Closed web server.
Terminating browser process..
emrun quitting with process exit code 0
```
Note the warning of emrun got deatched from browser process and also the fact that emrun tried to kill the browser process at exit but to no avail.

I will try to apply your patch and post my test result in another comment post when I am done.

**Comment by weitjong** [comment]
Sorry that I got a little sidetrack for applying your patch but I think I am beginning to figure out what is the root cause of the emrun failure to kill the browser process in the first place. In my test setup I had not used the ```--browser``` option before. I notice that without choosing a specific browser, emrun script actually spawns the browser process indirectly via [```xdg-open```](http://linux.die.net/man/1/xdg-open) (see the verbose log above). Similarly on my test Mac OS X host system, this setup spawns the browser process indirectly via [```open```](https://developer.apple.com/library/mac/documentation/Darwin/Reference/ManPages/man1/open.1.html). I suspect that this has somehow interfered with the emrun script to get the correct browser PID to kill. I have just now tested my suspicion by specifically choosing a browser id to run in my test setup. After the change, emrun spawns the browser process directly and kill at exit successfully!

Here is the snippet of a successful integration tests log with Chrome.
```
$ rake make target=test emscripten
Running tests...
Test project /home/weitjong/SDKs/urho3d/emscripten-Build
      Start  1: SpritesLua
 1/36 Test  #1: SpritesLua .......................   Passed   23.77 sec
      Start  2: 01_HelloWorld
 2/36 Test  #2: 01_HelloWorld ....................   Passed   21.90 sec
      Start  3: 02_HelloGUI
 3/36 Test  #3: 02_HelloGUI ......................   Passed   22.42 sec
      Start  4: 03_Sprites
 4/36 Test  #4: 03_Sprites .......................   Passed   22.53 sec
      Start  5: 04_StaticScene
 5/36 Test  #5: 04_StaticScene ...................   Passed   21.96 sec
      Start  6: 05_AnimatingScene
...
33/36 Test #33: 28_Urho2DPhysicsRope .............   Passed   21.89 sec
      Start 34: 32_Urho2DConstraints
34/36 Test #34: 32_Urho2DConstraints .............   Passed   22.39 sec
      Start 35: 33_Urho2DSpriterAnimation
35/36 Test #35: 33_Urho2DSpriterAnimation ........   Passed   22.40 sec
      Start 36: 36_Urho2DTileMap
36/36 Test #36: 36_Urho2DTileMap .................   Passed   22.28 sec

100% tests passed, 0 tests failed out of 36

Total Test time (real) = 804.98 sec
```

The integration tests failed to run all the way with Firefox. I have retried it a few times but it seems that after a few passed test cases then Firefox crash on the subsequent test case and a Firefox safe-mode dialog window pop up halting the whole integration tests process. I have not looked into the Emrun script in detail but perhaps Firefox does not like the way it is being killed/terminated in the script but I guess this is totally another issue.

I also notice another problem with the way emrun kills the browser process.
```
Terminating all processes that have string 'chrome' in their name.
```
If I have a chrome session already opened for other purpose and emrun spawns another one then it should only kill its own chrome process and leave my other session alone.

To summarize:
- The warning of emrun detached from browser process does not seem to have any bearing of the failure. The way how the browser process is spawned does.
- On the flip side, when it succeed to kill the browser process, it may actually kill the wrong one at exit.

**Comment by juj** [comment]
Oh, I see. That is completely correct. I will add an instruction text to suggest explicitly using `--browser` to avoid detached mode. That should avoid blocking on this issue in the future.

I remember having written code that only kills the PID that it spawned, however, I recall that there was a problem that it didn't actually work in multiprocess browsers. I should probably revisit that.

To set up Firefox for best `emrun` experience, navigate to `about:config` in Firefox, and set the following prefs:

	; Make sure to unblock popups being spawned from http://localhost/
	browser.popups.showPopupBlocker;false

	; Don't ask the user to change the default browser when spawning the browser
	browser.shell.checkDefaultBrowser;false

	; Don't autorestore previous tabs, just open the one from the command line
	browser.sessionstore.resume_from_crash;false
	services.sync.prefs.sync.browser.sessionstore.restore_on_demand;false
	browser.sessionstore.restore_on_demand;false

	; Don't bring up the modal "Start in Safe Mode" dialog after browser is killed, since we are doing automated runs
	browser.sessionstore.max_resumed_crashes;0
	toolkip.startup.max_resumed_crashes;0

	; Don't fail on long-running scripts, but have emrun instead control execution termination
	dom.max_script_run_time;0
	dom.max_chrome_script_run_time;0

	; (Optional) Accelerate browser update background timer tick so that autoupdates take place as quickly as possible and latest browser is always being tested.
	app.update.download.backgroundInterval;1

	; (Optional) Always run in private browsing mode to avoid caching any pages (but also disables IndexedDB persistency!)
	browser.privatebrowsing.autostart;true

I'll add these to `emrun` documentation itself, so that people will have better time finding them.

**Comment by weitjong** [comment]
Hi Jukka, thanks for your support. I have applied the Firefox configuration changes as you recommended with one minor change below. I have to override these two settings from 1 and 3 to 1000 and 1000, respectively, instead of 0 and 0.
```
; Don't bring up the modal "Start in Safe Mode" dialog after browser is killed, since we are doing automated runs
browser.sessionstore.max_resumed_crashes;1000
toolkip.startup.max_resumed_crashes;1000
```
Setting them to 0 would actually make it worse which brings the safe mode dialog immediately after the first passed test case.

With this change, the integration tests ran successfully with Firefox and it was completed slightly faster than Chrome with the same timeout setting.

```
33/36 Test #33: 28_Urho2DPhysicsRope .............   Passed   19.07 sec
      Start 34: 32_Urho2DConstraints
34/36 Test #34: 32_Urho2DConstraints .............   Passed   18.70 sec
      Start 35: 33_Urho2DSpriterAnimation
35/36 Test #35: 33_Urho2DSpriterAnimation ........   Passed   19.25 sec
      Start 36: 36_Urho2DTileMap
36/36 Test #36: 36_Urho2DTileMap .................   Passed   19.15 sec

100% tests passed, 0 tests failed out of 36

Total Test time (real) = 672.89 sec
```

I think we can consider this issue as close. Thanks again for your help.

**Comment by juj** [comment]
Oh sorry, try setting the value to -1 instead of zero. Does that disable
the popup?

Let's keep the issue open until I have committed the doc fixes.

**Comment by weitjong** [comment]
Yes, -1 works too.

**Comment by weitjong** [comment]
I got time to test out your patch. With the patch applied, the ```--timeout``` now works as expected (tested on Linux host system). I believe I have tried that emrun option earlier and found that it did not work before. So, I would say the patch works.

**Comment by juj** [comment]
Ok, pushed in the fixes.

As for the issue where emrun kills all processes by name instead of the specific browser process by PID, I think that should occur only in detached mode, and now that you are using `--browser`, it should no longer occur. Let me know if that is not the case. I'll close this as fixed now.
